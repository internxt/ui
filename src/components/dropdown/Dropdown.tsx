import { useState, ReactNode, useEffect, useRef } from 'react';
import { Menu, MenuItemType } from '../';

export type DropdownProps<T> = {
  children: ReactNode | ((obj: { open: boolean }) => JSX.Element);
  options?: { text: string; onClick: () => void }[];
  classButton?: string;
  menuItems?: ReactNode[];
  classMenuItems: string;
  openDirection: 'left' | 'right';
  dropdownActionsContext?: Array<MenuItemType<T>>;
  item?: T;
};

/**
 * Dropdown component
 *
 * @property {ReactNode | ((obj: { open: boolean }) => JSX.Element)} children
 * - The content of the dropdown button. It can be a React component or a function that receives an object
 * with the `open` property to render the dropdown's state.
 *
 * @property {Object[]} [options]
 * - The dropdown options, where each option contains a text and an `onClick` function
 * that is executed when the option is clicked.
 *
 * @property {string} [classButton]
 * - Additional classes for the dropdown button, allowing customization of its appearance.
 *
 * @property {ReactNode[]} [menuItems]
 * - Menu items to be added to the dropdown. These can be React components rendered inside the menu.
 *
 * @property {string} classMenuItems
 * - Additional CSS classes for the container of menu items, allowing for custom styling.
 *
 * @property {'left' | 'right'} openDirection
 * - The direction in which the dropdown menu opens. It can be 'left' or 'right'.
 *
 * @property { Array<MenuItemType<T>>} [dropdownActionsContext]
 * - Additional actions that can be passed to the dropdown menu.
 * Used for extending the menu with more options or functionalities.
 *
 * @property {T} [item]
 * - The current item that may be used within the options or actions of the menu,
 * allowing customization of the dropdown's logic.
 */

const Dropdown = <T,>({
  children,
  options,
  classButton,
  menuItems,
  classMenuItems,
  openDirection,
  dropdownActionsContext,
  item,
}: DropdownProps<T>): JSX.Element => {
  const [isOpen, setIsOpen] = useState(false);
  const direction = openDirection === 'left' ? 'origin-top-left' : 'origin-top-right';
  const containerRef = useRef<HTMLDivElement>(null);

  useEffect(() => {
    const handleClickOutside = (e: MouseEvent) => {
      if (containerRef.current && !containerRef.current.contains(e.target as Node)) {
        setIsOpen(false);
      }
    };

    document.addEventListener('mousedown', handleClickOutside);
    document.addEventListener('contextmenu', handleClickOutside);

    return () => {
      document.removeEventListener('mousedown', handleClickOutside);
      document.removeEventListener('contextmenu', handleClickOutside);
    };
  }, []);

  const group1: Array<MenuItemType<T>> = options
    ? options.map((option) => ({
        name: option.text,
        action: () => option.onClick(),
      }))
    : [];

  const group2: Array<MenuItemType<T>> = menuItems
    ? menuItems.map((menuItem) => ({
        node: menuItem,
      }))
    : [];

  const group3: Array<MenuItemType<T>> = dropdownActionsContext || [];

  const allItems = [...group1, ...group2, ...group3];

  const toggleMenu = () => setIsOpen((prev) => !prev);
  const closeMenu = () => setIsOpen(false);

  return (
    <div className="relative outline-none" ref={containerRef}>
      <button
        className={`cursor-pointer outline-none ${classButton}`}
        onClick={toggleMenu}
        aria-expanded={isOpen}
        aria-haspopup="menu"
      >
        {typeof children === 'function' ? children({ open: isOpen }) : children}
      </button>

      <div
        className={`absolute ${openDirection === 'left' ? 'left-0' : 'right-0'} 
          transform shadow-subtle-hard transition-all duration-100 ease-in-out ${
            isOpen ? `scale-100 opacity-100 ${direction}` : 'pointer-events-none scale-95 opacity-0'
          }`}
        data-testid="menu-dropdown"
      >
        <div className={`absolute ${classMenuItems}`}>
          <Menu item={item} isOpen={isOpen} handleMenuClose={closeMenu} menu={allItems} />
        </div>
      </div>
    </div>
  );
};

export default Dropdown;
