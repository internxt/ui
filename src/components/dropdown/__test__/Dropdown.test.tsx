import React from 'react';
import { render, fireEvent, waitFor } from '@testing-library/react';
import { describe, it, expect, vi, afterEach } from 'vitest';
import Dropdown from '../Dropdown';

describe('Dropdown component', () => {
  const handleClick1 = vi.fn();
  const handleClick2 = vi.fn();
  const options = [
    { text: 'Option 1', onClick: handleClick1, disabled: () => false },
    { text: 'Option 2', onClick: handleClick2, disabled: () => true },
  ];
  const mockAction = vi.fn();
  const dropdownActionsContext = [
    {
      name: 'Action 1',
      action: mockAction,
      separator: false,
    },
  ];
  const renderDropdown = () => {
    return render(
      <Dropdown
        openDirection="left"
        classMenuItems="menu"
        classButton="button"
        options={options}
        dropdownActionsContext={dropdownActionsContext}
        item={{}}
      >
        Dropdown
      </Dropdown>,
    );
  };

  afterEach(() => {
    vi.clearAllMocks();
  });

  it('should match snapshot when isOpen is true', () => {
    const { container } = renderDropdown();
    expect(container).toMatchSnapshot();
  });

  it('should toggle menu open and close on button click', async () => {
    const { getByTestId, getByRole } = renderDropdown();

    const button = getByRole('button');
    const menu = getByTestId('menu-dropdown');

    expect(menu).toHaveClass('opacity-0');
    expect(menu).toHaveClass('scale-95');

    fireEvent.click(button);

    await waitFor(() => {
      expect(menu).not.toHaveClass('opacity-0');
      expect(menu).not.toHaveClass('scale-95');
      expect(menu).toHaveClass('opacity-100');
      expect(menu).toHaveClass('scale-100');
    });

    fireEvent.click(button);

    await waitFor(() => {
      expect(menu).toHaveClass('opacity-0');
      expect(menu).toHaveClass('scale-95');
    });
  });

  it('should render menu options correctly', () => {
    const { getByText, getByRole } = renderDropdown();

    fireEvent.click(getByRole('button'));

    expect(getByText('Option 1')).toBeInTheDocument();
    expect(getByText('Option 2')).toBeInTheDocument();
  });

  it('should call onClick when an option is clicked', async () => {
    const { getByRole, getByText } = renderDropdown();

    fireEvent.click(getByRole('button'));
    fireEvent.click(getByText('Option 1'));

    await waitFor(() => expect(handleClick1).toHaveBeenCalledTimes(1));
  });

  it('should close the menu after selecting an option', async () => {
    const { getByRole, getByText, getByTestId } = renderDropdown();

    fireEvent.click(getByRole('button'));
    fireEvent.click(getByText('Option 1'));

    const menu = getByTestId('menu-dropdown');

    await waitFor(() => {
      expect(menu).not.toHaveClass('opacity-100');
      expect(menu).not.toHaveClass('scale-100');
      expect(menu).toHaveClass('opacity-0');
      expect(menu).toHaveClass('scale-95');
    });
  });

  it('should render dropdownActionsContext correctly', async () => {
    const { getByRole, getByText } = renderDropdown();

    fireEvent.click(getByRole('button'));

    expect(getByText('Action 1')).toBeInTheDocument();

    fireEvent.click(getByText('Action 1'));

    await waitFor(() => expect(mockAction).toHaveBeenCalledTimes(1));
  });

  it('should navigate through items with ArrowDown and ArrowUp keys', async () => {
    const { getByRole, getByText } = renderDropdown();

    fireEvent.click(getByRole('button'));

    fireEvent.keyDown(window, { key: 'ArrowDown' });

    const firstOption = getByText('Option 1');
    const secondOption = getByText('Option 2');

    expect(firstOption).toHaveClass('bg-gray-5 text-gray-100');
    expect(secondOption).not.toHaveClass('bg-gray-5 text-gray-100');

    fireEvent.keyDown(window, { key: 'ArrowDown' });
    expect(firstOption).not.toHaveClass('bg-gray-5 text-gray-100');
    expect(secondOption).toHaveClass('bg-gray-5 text-gray-100');

    fireEvent.keyDown(window, { key: 'ArrowUp' });
    expect(firstOption).toHaveClass('bg-gray-5 text-gray-100');
    expect(secondOption).not.toHaveClass('bg-gray-5 text-gray-100');
  });

  it('should trigger the onClick handler when Enter is pressed on an item', async () => {
    const { getByRole } = renderDropdown();

    fireEvent.click(getByRole('button'));
    fireEvent.keyDown(window, { key: 'ArrowDown' });
    fireEvent.keyDown(window, { key: 'Enter' });

    expect(handleClick1).toHaveBeenCalledTimes(1);
  });

  it('should close the menu after selecting an option with Enter', async () => {
    const { getByRole, getByTestId } = renderDropdown();

    fireEvent.click(getByRole('button'));
    fireEvent.keyDown(window, { key: 'ArrowDown' });
    fireEvent.keyDown(window, { key: 'Enter' });

    const menu = getByTestId('menu-dropdown');
    expect(menu).toHaveClass('scale-95 opacity-0');
  });
});
