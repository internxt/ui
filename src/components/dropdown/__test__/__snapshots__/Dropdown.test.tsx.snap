// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Dropdown component > should match snapshot 1`] = `
<div>
  <div
    class="relative outline-none"
  >
    <button
      aria-expanded="false"
      aria-haspopup="menu"
      class="cursor-pointer outline-none button"
    >
      Dropdown
    </button>
    <div
      class="absolute left-0 
          transform shadow-subtle-hard transition-all duration-100 ease-in-out pointer-events-none scale-95 opacity-0"
      data-testid="menu-dropdown"
    >
      <div
        class="absolute menu"
      >
        <div
          class="z-20 mt-0 flex flex-col rounded-lg bg-surface py-1.5 shadow-subtle-hard outline-none dark:bg-gray-5"
        >
          <div>
            <div
              class=""
            >
              <div
                class="flex cursor-pointer flex-row whitespace-nowrap px-4 py-1.5 text-base
                    
                    
                    
                    text-gray-80
                  "
              >
                <div
                  class="flex flex-row items-center space-x-2"
                >
                  <span>
                    Option 1
                  </span>
                </div>
                <span
                  class="ml-5 flex grow items-center justify-end text-sm text-gray-40"
                />
              </div>
            </div>
          </div>
          <div>
            <div
              class=""
            >
              <div
                class="flex cursor-pointer flex-row whitespace-nowrap px-4 py-1.5 text-base
                    
                    
                    
                    text-gray-80
                  "
              >
                <div
                  class="flex flex-row items-center space-x-2"
                >
                  <span>
                    Option 2
                  </span>
                </div>
                <span
                  class="ml-5 flex grow items-center justify-end text-sm text-gray-40"
                />
              </div>
            </div>
          </div>
          <div>
            <div
              class=""
            >
              <div
                class="flex cursor-pointer flex-row whitespace-nowrap px-4 py-1.5 text-base
                    
                    
                    
                    text-gray-80
                  "
              >
                <div
                  class="flex flex-row items-center space-x-2"
                >
                  <span>
                    Action 1
                  </span>
                </div>
                <span
                  class="ml-5 flex grow items-center justify-end text-sm text-gray-40"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
